File: 54-spiral-matrix/spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int r = matrix.size();
        int c = matrix[0].size();
        int top = 0, bottom = r, left = 0, right = c;
        vector<int> result;
        while(top<bottom and left<right){
            //Moving towards right direction:
            for(int i = left; i<right; i++){
                result.push_back(matrix[left][i]);
            }
            top++;

            

            //Moving down:
            for(int i = top; i<bottom; i++){
                result.push_back(matrix[i][right-1]);
            }
            right--;

            if(top>=bottom || left>=right) break;

            //Moving left:
            for(int i = right-1; i>=left; i--){
                result.push_back(matrix[bottom-1][i]);
            }
            bottom--;

            //Moving up:
            for(int i = bottom-1; i>=top; i--){
                result.push_back(matrix[i][left]);
            }
            left++;
        }
        return result;
    }
};

File: 54-spiral-matrix/spiral-matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        result = []
        top, bottom = 0, len(matrix)
        left, right = 0, len(matrix[0])

        while left < right and top < bottom:
            # Traverse from Left to Right
            i = left
            while i < right:
                result.append(matrix[top][i])
                i += 1
            top += 1

            # Traverse from Top to Bottom
            i = top
            while i < bottom:
                result.append(matrix[i][right - 1])
                i += 1
            right -= 1

            # Check if we still have rows and columns left
            if not (left < right and top < bottom):
                break

            # Traverse from Right to Left
            i = right - 1
            while i >= left:
                result.append(matrix[bottom - 1][i])
                i -= 1
            bottom -= 1

            # Traverse from Bottom to Top
            i = bottom - 1
            while i >= top:
                result.append(matrix[i][left])
                i -= 1
            left += 1

        return result
